### DISCUSSION TITLE
Is `splitChunks.automaticNameDelimiter` still used in webpack 5?

### DISCUSSION THREAD

**Discussion by scottohara** [webpack/webpack#14451] [Q&A]
Originally asked at https://stackoverflow.com/questions/69490911/webpack-5-shared-chunk-file-name

In short:
* In v4, a chunk containing shared code from `admin.js` and `app.js` chunks was previously named `admin~app.js`
* In v5 production mode, the same shared chunk has a numeric name (e.g. `121.js`); and in development mode it has a name using underscores (e.g. `some_long_name_involving_paths_to_the_original_source.js`)

Explicitly setting `automaticNameDelimiter: '~'` appears to have no effect.

Is this option now deprecated, or have I misunderstood its purpose?

**Comment by alexander-akait** [comment]
`automaticNameDelimiter` should work https://github.com/webpack/webpack/blob/d106f00db56e5b9a1c2ed43266f1837f846f4ce8/lib/optimize/SplitChunksPlugin.js#L649, maybe you override somewhere it? Can you provide example?

**Reply by scottohara** [reply]
Thanks @alexander-akait, appreciate your assistance.

I can provide a full example if needed, but here are the relevant parts:

```js
entry: {
  app: "app",
  admin: "admin"
},
optimization: {
  splitChunks: {
    chunks: "all",
    cacheGroups: {
      vendor: {
        name: "vendor",
        test: /[\\/]node_modules[\\/]/
      }
    }
  },
  runtimeChunk: "single"
}
```

With v5, `mode: "production"`, this config produces:

```shell
/dist/121.js        // shared chunk for app + admin
/dist/admin.js      // admin chunk
/dist/app.js        // app chunk
/dist/runtime.js    // runtime chunk
/dist/vendor.js     // vendor chunk
```

Note the shared chunk file name `121.js`.

With v5, `mode: "development"`, it produces:

```shell
/dist/src_main_frontend_errors.ts.js        // shared chunk for app + admin
/dist/admin.js      // admin chunk
/dist/app.js        // app chunk
/dist/runtime.js    // runtime chunk
/dist/vendor.js     // vendor chunk
```

Note the shared chunk file name `src_main_frontend_errors.ts.js` (the source was `/src/main/frontend/errors.ts`).



In contrast, in v4 the config was:

```js
entry: {
  app: "app",
  admin: "admin"
},
output: {
  filename: "[name].js"
},
optimization: {
  moduleIds: "hashed",
  splitChunks: {
    chunks: "all",
    cacheGroups: {
      vendor: {
        name: "vendor",
        test: /[\\/]node_modules[\\/]/
      }
    }
  },
  runtimeChunk: "single"
}
```

(includes `output.filename` and `optimization.moduleIds`, which are now defaults in v5)

This previously produced:

```shell
/dist/admin.js      // admin chunk
/dist/admin~app.js  // shared chunk for app + admin
/dist/app.js        // app chunk
/dist/runtime.js    // webpack runtime chunk (from runtimeChunk: "single")
/dist/vendor.js     // vendor chunk (node_modules)
```

(Note the shared chunk file name `admin~app.js`)

**Reply by alexander-akait** [reply]
> Note the shared chunk file name 121.js.

Because we implement long term caching out of box and minify your name to reduce size of bundle, I strongly recommend to keep default options, less name = less bytes = faster load.

> /dist/src_main_frontend_errors.ts.js        // shared chunk for app + admin

Are you sure you provided full configuration?

Note - webpack by default have `vendor` and `default` groups (they should be used in 98% cases, I recommend to use them) https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1070, you should disable them if you need custom splitting, note - you can decrease perf of your application if you're not careful with these options

**Reply by scottohara** [reply]
Thanks @alexander-akait .

> Are you sure you provided full configuration?

No, sorry I had only provided what I thought were the relevant parts.

Here is a complete (reduced) example that may help to demonstrate my question:


```js
module.exports = {
  mode: 'development',
  entry: {
    a: './src/a.js',
    b: './src/b.js'
  },
  optimization: {
    splitChunks: {
      automaticNameDelimiter: '%',
      chunks: 'all',
      minSize: 0  // NOT RECOMMENDED!!
    }
  }
}
```

I appreciate that the `optimization` defaults are specifically chosen and should be used where possible.

I'm setting `splitChunks.minSize: 0` here only to force webpack to create more files in the output so that we can compare the differences between v4 and v5.

I'm also setting `splitChunks.automaticNameDelimiter: '%'` so we can observe how it impacts the output file names.

My entrypoints `a.js` & `b.js` both reference a shared module `c.js`:

```js
// a.js
import c from "./c.js";
c();

// b.js
import c from "./c.js";
c();

// c.js
export default function() {
	return "c";
}
```

In v4, the output in `/dist` is:

```shell
a.js
a%b.js    // without automaticNameDelimiter set this would normally be a~b.js
b.js
```

In v5, the output in `/dist` is:

```shell
a.js
b.js
src_c_js.js    // name appears to be derived from the path of the original file (/src/c.js)
```

So my question is: if `splitChunks.automaticNameDelimiter` no longer appears in the name of the shared chunk in v5, when is it used?

I hope this helps clarify my question.

**Reply by alexander-akait** [reply]
Please provide full configuration or steps to reproduce, `automaticNameDelimiter` should be used, we have tests on it, so we need investigate what is wrong...

**Reply by scottohara** [reply]
Sorry @alexander-akait , I’m not sure what you mean by “full configuration or steps to reproduce”.

The above shows the config and steps. Do you need me to create a repo with the above files that you can clone?

**Reply by alexander-akait** [reply]
Yep, thanks

**Reply by scottohara** [reply]
Example repo with reproduction steps is here: https://github.com/scottohara/webpack5-automaticnamedelimiter

1. Checkout `v4` branch, run `npm install`, run `webpack`, inspect `/dist`
2. Checkout `v5` branch, run `npm install`, run `webpack`, inspect `/dist`

I hope this is sufficient for further investigation.

**Reply by alexander-akait** [reply]
Here's an interesting situation - webpack v5 by default split shared modules for multiple entries, so you just have `a`, `b`, and unnamed chunks (shared chunk), split chunk just enforce creation of shared chunk, but name doesn't exist, for shared chunk we use `id` property (i.e. `src_c_js` in your case) and than webpack uses `id` to generate name, so `automaticNameDelimiter` just not used here, this can be a little confusing, but from a code point of view it is pretty logical, we don't have name, so we can't apply `automaticNameDelimiter` :smile: Anyway we can force it:
```
minSize: 0,
filename: (pathData) => {
  // Let's give name for unnamed chunks, you can achieve the same using `webpackChunkName` comment 
  if (!pathData.chunk.name) {
    return `${pathData.chunk.id.replace(/_/g, "%")}.js`;
  }
				
  return "[name].js";
},
```

**Reply by alexander-akait** [reply]
By default we use `_` for concat modules in shared chunks

**Reply by ianjosephwilson** [reply]
I've been trying to follow this issue because it relates to a problem I have.  How will an app know which chunks to load for an entry point if the entry point names are not in the shared chunk filenames ?  (and the app is not using the html plugin)  Ie. If the name is `121.js` how does app know that file should be loaded before `a.js`?

**Reply by alexander-akait** [reply]
@ianjosephwilson from chunk graph, you can found it in `stats`
